core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -628,6 +628,11 @@ private[spark] class TaskSetManager(
     // here "result.value()" just returns the value and won't block other threads.
     sched.dagScheduler.taskEnded(
       tasks(index), Success, result.value(), result.accumUpdates, info, result.metrics)
+    // kill other attempts of this task when speculation is enabled.
```

Comment:
I am not so sure about this part, but will that be better if we moved the code after we confirmed the task succeeded?